apps/remix-ide/src/app/tabs/hardhat-provider.js

```diff
@@ -32,19 +32,30 @@ export default class HardhatProvider extends Plugin {
         <br>
         For more info, visit: <a href="https://hardhat.org/getting-started/#connecting-a-wallet-or-dapp-to-hardhat-network" target="_blank">Hardhat Documentation</a>
         <br><br>
+        <b>Note:</b> Click on 'Cancel' if node is stopped.
+        <br><br>
         Hardhat JSON-RPC Endpoint
       </div>
     `
   }
 
   sendAsync (data) {
     return new Promise((resolve, reject) => {
-      if (!this.provider) {
+      // If provider is not set, allow to open modal only when provider is trying to connect
+      if (!this.provider || data.method === 'net_listening') {
         modalDialogCustom.prompt('Hardhat node request', this.hardhatProviderDialogBody(), 'http://127.0.0.1:8545', (target) => {
           this.provider = new Web3.providers.HttpProvider(target)
           this.sendAsyncInternal(data, resolve, reject)
         }, () => {
-          this.sendAsyncInternal(data, resolve, reject)
+          // If 'cancel' is clicked while trying to connect, handle it in custom manner
+          if (data.method === 'net_listening') resolve({ jsonrpc: '2.0', result: 'canceled', id: data.id })
+          else {
+            // When node is abruptly stopped, modal will appear
+            // On which clicking on 'Cancel' will set the Envrionment to VM
+            this.blockchain.changeExecutionContext('vm')
+            this.provider = this.blockchain.getCurrentProvider()
```

Comment:
so `this.provider` will contain the VM provider in that case?

Reply:
if there's no connection, i'd just reset `this.provider` in that case.

Reply:
That is the issue, it opens modal again if `(!this.provider || data.method === 'net_listening')` returns true. with VM provider, it will go to `else` and in `sendAsyncInternal ` , it will throw error on failing `(this.blockchain.getProvider() !== 'Hardhat Provider' && data.method !== 'net_listening')` condition.

Reply:
ok maybe set it to a fake provider (object which has an empty sendAsync method), because setting it to the current provider is note really accurate.

Reply:
We set it to VM when user clicks on `cancel`